File: dotnet/Metrics/GenresCombinations.cs
namespace Southworks.Demos.Spark.Metrics
{
    using Microsoft.Spark.Sql;
    using static Microsoft.Spark.Sql.Functions;

    public class GenresCombinations : IMetricsSolver
    {
        public DataFrame Run(DataFrame moviesDataFrame, DataFrame ratingsDataFrame)
        {
            // Get a genres list.
            var firstGenresList = moviesDataFrame
                .WithColumn("genres", Split(Col("genres"), "\\|"))
                .Select(Col("movieId"), Explode(Col("genres")))
                .WithColumnRenamed("col", "genre");

            // Get a second list of genres with different column names for same-table join.
            var secondGenresList = firstGenresList
                .WithColumnRenamed("genre", "r_genre")
                .WithColumnRenamed("movieId", "r_movieId");

            // Join both genre lists to get every combination. Group by both genres in the combination. Count repetitions and filter the greatest for each one.
            var greatestCombinationByGenre = firstGenresList
                .Join(secondGenresList, firstGenresList.Col("genre").NotEqual(secondGenresList.Col("r_genre")))
                .Where(firstGenresList.Col("movieId").EqualTo(secondGenresList.Col("r_movieId")))
                .GroupBy("genre", "r_genre")
                .Count()
                .OrderBy(Col("count").Desc())
                .GroupBy("genre")
                .Agg(First("r_genre").As("most related genre"), Max("count").As("times related"))
                .OrderBy("genre")
                .Select("times related", "genre", "most related genre");

            return greatestCombinationByGenre;
        }
    }
}

File: java/src/main/java/com/southworks/demos/metrics/implementation/GenresCombinations.java
package com.southworks.demos.metrics.implementation;

import com.southworks.demos.metrics.MetricsSolver;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;

import static org.apache.spark.sql.functions.*;

public class GenresCombinations implements MetricsSolver {

    @Override
    public Dataset<Row> run(Dataset<Row> moviesDataset, Dataset<Row> ratingsDataset) {

        // Get a genres list.
        Dataset<Row> firstGenresList = moviesDataset
                .withColumn("genres", split(col("genres"), "\\|"))
                .select(col("movieId"), explode(col("genres")))
                .withColumnRenamed("col", "genre");

        // Get a second list of genres with different column names for same-table join.
        Dataset<Row> secondGenresList = firstGenresList
                .withColumnRenamed("genre", "r_genre")
                .withColumnRenamed("movieId", "r_movieId");

        // Join both genre lists to get every combination. Group by both genres in the combination. Count repetitions and filter the greatest for each one.
        Dataset<Row> greatestCombinationByGenre = firstGenresList
                .join(secondGenresList, firstGenresList.col("genre").notEqual(secondGenresList.col("r_genre")))
                .where(firstGenresList.col("movieId").equalTo(secondGenresList.col("r_movieId")))
                .groupBy("genre", "r_genre")
                .count()
                .orderBy(col("count").desc())
                .groupBy("genre")
                .agg(first("r_genre").as("most related genre"),
                        max("count").as("times related"))
                .orderBy("genre")
                .select("times related", "genre", "most related genre");

        return greatestCombinationByGenre;
    }
}
